File: numbers/src/main/java/mathtools/numbers/factors/IntOperations.java
package mathtools.numbers.factors;

import com.google.common.math.IntMath;

import javax.annotation.Nonnull;

import mathtools.pairs.IntPair;

/** Functions that operate on 32-bit Integers
 * @author DK96-OS : 2020 - 2022 */
public final class IntOperations {

	/** Perform Base 10 shift up to Int Max value.
	 *  If an overflow occurs, Int MaxValue is returned
	 *  Negative shifts are allowed
	 * @param x The value to shift
	 * @param e The power of base 10 applied
	 * @return The shifted value, or MaxValue if known that the shift is too large */
	public static int tenShift(
		final int x,
		final int e
	) {
		if (0 == e)
			return x;
		else if (0 == x)
			return 0;
		else if (0 > x)
			return -tenShift(-x, e);
		else if (9 < e)
			return Integer.MAX_VALUE;
		else if (0 > e) {
			if (-9 > e) return 0;
			else return
				(-3 > e ? tenShift(x / 10000, e + 4) :
					(-3 == e ? x / 1000 :
						(-2 == e ? x / 100 : x / 10)
					)
				);
		} else {
			final long testShift;
			switch(e) {
				case 1:
					testShift = (long)x * 10L;
					break;
				case 2:
					testShift = (long)x * 100L;
					break;
				case 3:
					testShift = (long)x * 1000L;
					break;
				default:
					testShift = (long)x * 10000L;
			}
			if ((long) Integer.MAX_VALUE >= testShift)
				return (5 > e ? (int)testShift :
					tenShift((int)testShift, e - 4)
				);
			else
				return Integer.MAX_VALUE;
		}
	}

	/** Tries to evaluate an exponential term within 32-bit signed int
	 *  If the term cannot be evaluated, the number of powers remaining are returned as well
	 * @param base The base factor of the exponent.
	 * @param power The power of the exponent.
	 * @return A pair of integers. If the second int is zero, the first is the evaluated term. If
	the second int is non-zero, this is the power on the exponent that has yet to be evaluated. * */
	@Nonnull
	public static IntPair exponent(
		final int base,
		int power
	) {
		if (0 > power)
			return new IntPair(base, power);
		else if (0 == power)
			return new IntPair(1, 0);
		else if (2 > base)
			return negativeExponentBase(base, power);
		else if (1 == power)
			return new IntPair(base, 0);
		else {
			int product = base;
			power--;
			try {
				product = IntMath.checkedMultiply(product, base);
				power--;
				for (; 0 < power; --power) {
					product = IntMath.checkedMultiply(product, base);
				}
				return new IntPair(product, 0);
			} catch (final ArithmeticException e) {
				return new IntPair(product, power);
			}
		}
	}

	/** Handles certain exponent function cases
	 * @param x The base factor of the exponent. Is 1, 0, or negative.
	 * @param power The power of the exponent. Is greater than 0.
	 * @return A pair of integer values */
	@Nonnull
	private static IntPair negativeExponentBase(
		final int x,
		final int power
	) {
		switch (x) {
			case -1: if (BitFactoring.isProductOf2(power))
				return new IntPair(1, 0);
			else
				return new IntPair(-1, 0);
			case 0:
				return new IntPair(0, 0);
			case 1:
				return new IntPair(1, 0);
			default:
				// X is negative, need to flip
				final IntPair result = exponent(-x, power);
				// Check the results for additional processing
				int product = result.getFirst();
				final int remaining = result.getSecond();
				// If the difference is odd, the product will be negative
				if (!BitFactoring.isProductOf2(power - remaining))
					product = -product;
				return new IntPair(product, remaining);
		}
	}

	private IntOperations() {}

}

File: numbers/src/main/kotlin/mathtools/numbers/factors/IntOperations.kt
package mathtools.numbers.factors

import mathtools.numbers.factors.NumberFactors.isProductOf2

/** Functions that operate on 32-bit Integers
 * @author DK96-OS : 2020 - 2022 */
object IntOperations {

	/** Perform Base 10 shift up to Int Max value.
	 *  If an overflow occurs, Int MaxValue is returned
	 *  Negative shifts are allowed
	 * @param x The value to shift
	 * @param e The power of base 10 applied
	 * @return The shifted value, or MaxValue if known that the shift is too large */
	fun tenShift(
		x: Int,
		e: Int,
	) : Int = when {
		e == 0 -> x
		x == 0 -> 0
		x < 0 -> -tenShift(-x, e)
		e < 0 -> when {
			e < -9 -> 0
			e < -3 -> tenShift(x / 10_000, e + 4)
			e == -3 -> x / 1000
			e == -2 -> x / 100
			else -> x / 10
		}
		e > 9 -> Int.MAX_VALUE
		else -> {
			// Protect against overflow using Long
			val testShift = when (e) {
				1 -> x * 10L
				2 -> x * 100L
				3 -> x * 1000L
				else -> x * 10_000L
			}
			if (testShift <= Int.MAX_VALUE) {
				if (e < 5) testShift.toInt()
				else tenShift(testShift.toInt(), e - 4)
			} else Int.MAX_VALUE
		}
	}

	/** Compute an exponent, while checking for Integer Overflow
	 * Negative Powers are not allowed - will be returned
	 * @param x The base value of the exponent
	 * @param power The power of the exponent
	 * @return The product, and the remaining power that would cause an overflow */
	fun exponent(
		x: Int,
		power: Int,
	) : Pair<Int, Int> {
		when {
			power < 0 -> return x to power
			power == 0 -> return 1 to 0
			x < 2 -> return when (x) {
				1 -> 1 to 0
				0 -> 0 to 0
				-1 -> when {
					isProductOf2(power) -> 1 to 0
					else -> -1 to 0
				}
				else -> {
					val (product, remaining) = exponent(-x, power)
					if (isProductOf2(power - remaining))
						product to remaining
					else
						-product to remaining
				}
			}
			power == 1 -> return x to 0
		}
		val longX = x.toLong()
		var product = longX * longX
		// check for overflow
		if (product <= x || Int.MAX_VALUE < product)
			return x to power
		if (power == 2)
			return product.toInt() to 0
		for (e in 2 until power) {
			val next = product * longX
			if (next <= x || Int.MAX_VALUE < next) {
				// Integer Overflow
				return product.toInt() to power - e
			}
			product = next
		}
		return product.toInt() to 0
	}

}